prboom2/src/MUSIC/libopenmptplayer.c

```diff
@@ -0,0 +1,195 @@
+ *  of the License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+ *  02111-1307, USA.
+ *
+ * DESCRIPTION:
+ *  Tracker music player based on libopenmpt.
+ *---------------------------------------------------------------------
+ */
+
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "musicplayer.h"
+
+#ifndef HAVE_LIBOPENMPT
+#include <stddef.h>
+
+static const char *mpt_name (void)
+{
+  return "libopenmpt tracker player (DISABLED)";
+}
+
+
+static int mpt_init (int samplerate)
+{
+  return 0;
+}
+
+const music_player_t mpt_player =
+{
+  mpt_name,
+  mpt_init,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL
+};
+
+#else // HAVE_LIBOPENMPT
+
+#include <libopenmpt/libopenmpt.h>
+#include <string.h>
+#include <stdlib.h>
+#include <math.h>
+#include "lprintf.h"
+
+
+static int mpt_samplerate;
+static int mpt_paused;
+static int mpt_gain;
+static openmpt_module *mpt_cur_module;
+
+static const char *mpt_name (void)
+{
+  return "libopenmpt tracker player";
+}
+
+
+static int mpt_init (int samplerate)
+{
+  mpt_samplerate = samplerate;
+
+  return 1;
+}
+
+static void mpt_shutdown (void)
+{
+}
+
+static void mpt_setvolume (int v)
+{
+  int ret = 1;
+
+  // For 0% volume, log10f(0.f) yields -inf, which after converting to
+  // int yields INT_MIN gain.
+  mpt_gain = 2000.f * log10f(v / 15.f);
+
+  if (mpt_cur_module)
+    ret = openmpt_module_set_render_param(mpt_cur_module,
+                    OPENMPT_MODULE_RENDER_MASTERGAIN_MILLIBEL,
+                    mpt_gain);
+
+  if (!ret)
+    lprintf(LO_WARN, "mpt_setvolume: failed to set current module master gain\n");
+}
+
+static const void* mpt_registersong (const void *data, unsigned len)
+{
+  openmpt_module *mod = NULL;
+
+  mod = openmpt_module_create_from_memory2(data, len, NULL, NULL, NULL, NULL, NULL, NULL, NULL);
+  if (!mod)
+    lprintf(LO_WARN, "mpt_registersong: failed to open song\n");
+
+  return mod;
+}
+
+static void mpt_unregistersong (const void *handle)
+{
+  #pragma GCC diagnostic push
+  #pragma GCC diagnostic ignored "-Wcast-qual"
+  openmpt_module_destroy((openmpt_module *) handle);
+  #pragma GCC diagnostic pop
+}
+
+static void mpt_play (const void *handle, int looping)
+{
+  int ret;
+
+  #pragma GCC diagnostic push
+  #pragma GCC diagnostic ignored "-Wcast-qual"
+  mpt_cur_module = (openmpt_module *) handle;
+  #pragma GCC diagnostic pop
+  ret = openmpt_module_set_repeat_count(mpt_cur_module, looping ? -1 : 0);
+
+  if (!ret)
+    mpt_cur_module = NULL;
```

Comment:
We should return here if `openmpt_module_set_repeat_count` fails. Perhaps we shouldn't check for this error since it's unlikely to fail.